### PR TITLE
Allow compilation of build scripts against a different Gradle API version

### PR DESCRIPTION
**Pull request by wolfs**
This PRs explores what would be necessary to compile build scripts against a different Gradle API. It allows passing  the Gradle API version via `-Dorg.gradle.api.version=6.9` to the Gradle installation. Note that you also need a repository which contains an artifact for the Gradle API of this version with the coordinates `org.gradle:gradle-api:<version>`. You can specify the URL of the repository via `-Dgradle.api.repository.url=...`.

Currently this works for Kotlin and Groovy build scripts. It also should work for Kotlin pre-compiled scripts and Kotlin and Groovy build logic. Though it doesn't work for Groovy pre-compiled build scripts. I suppose we should be able to make this work with Groovy pre-compiled build scripts as well, though I did run into some classloader problems there.

You can also declare a dependency in your build script on the Gradle API with a version by using `gradleApi(<version>)` instead of `gradleApi()`.

Exploration doc: https://docs.google.com/document/d/1JB75X_Lt5BlubszqD0l10rHPqL89B2yd-yD39qizyDA/edit

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java [hunk 1]
```diff
@@ -383,6 +383,13 @@
      */
     Dependency gradleApi();
 
+    /**
+     * Creates a dependency on the API of the current version of Gradle.
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java:387]:
```suggestion
     * Creates a dependency on the API of the specified version of Gradle.
```

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java [hunk 2]
```diff
@@ -383,6 +383,13 @@
...
+     *
+     * @return The dependency.
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java:389]: It would be good to add some Javadoc on the expected format of the parameter. Are snapshots accepted?

I think we need to mention that only older versions are accepted, right? What happens if I give it a newer or non-existent version?

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java [hunk 3]
```diff
@@ -416,6 +425,44 @@ protected ITaskFactory createITaskFactory(Instantiator instantiator, TaskClassIn
...
+            fileResolver,
+            fileCollectionFactory,
+            dependencyMetaDataProvider,
+            new UnknownProjectFinder("Some unknown project"),
+            RootScriptDomainObjectContext.PLUGINS
+        );
+        return new GroovyScriptClasspathProvider(
+            classLoaderScopeRegistry.getCoreAndPluginsScope(),
+            () -> {
+                String version = System.getProperty("org.gradle.api.version");
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:446]: We should qualify what API version this is. I'd go with `org.gradle.api.target-version` or something.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java [hunk 4]
```diff
@@ -152,7 +152,17 @@ private void registerPlugins(Project project, GradlePluginDevelopmentExtension e
 
     private void applyDependencies(Project project) {
         DependencyHandler dependencies = project.getDependencies();
-        dependencies.add(API_CONFIGURATION, dependencies.gradleApi());
+        String apiVersion = System.getProperty("org.gradle.api.version");
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java:155]: We query this property in six different places by name. Instead it would be good to have a shared mechanism to expose this as an `Optional<VersionNumber>` instead. An `@Inject`-able service would be nice.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java [hunk 5]
```diff
@@ -152,7 +152,17 @@ private void registerPlugins(Project project, GradlePluginDevelopmentExtension e
...
+        if (apiVersion == null || apiVersion.isEmpty()) {
+            dependencies.add(API_CONFIGURATION, dependencies.gradleApi());
+        } else {
+            String repositoryUrl = System.getProperty("gradle.api.repository.url", "https://repo.gradle.org/gradle/libs-releases");
+            project.getRepositories().maven(repository -> {
+                repository.setName("gradleApiRepository");
+                repository.setUrl(repositoryUrl);
+            });
+            dependencies.add(API_CONFIGURATION, dependencies.gradleApi(apiVersion));
+        }
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java:165]: I think it would be good if the user could declare the API version they want to use via a `gradleApi("6.9")` call in the build script of the plugin. The above mechanism does not allow that.

File: subprojects/plugins/src/main/java/org/gradle/initialization/buildsrc/GroovyBuildSrcProjectConfigurationAction.java [hunk 6]
```diff
@@ -30,7 +30,17 @@ public void execute(ProjectInternal project) {
...
+        if (apiVersion == null || apiVersion.isEmpty()) {
+            dependencies.add(JavaPlugin.API_CONFIGURATION_NAME, dependencies.gradleApi());
+        } else {
+            String repositoryUrl = System.getProperty("gradle.api.repository.url", "https://repo.gradle.org/gradle/libs-releases");
+            project.getRepositories().maven(repository -> {
+                repository.setName("gradleApiRepository");
+                repository.setUrl(repositoryUrl);
+            });
+            dependencies.add(JavaPlugin.API_CONFIGURATION_NAME, dependencies.gradleApi(apiVersion));
+        }
```
Comment (lptr) [inline subprojects/plugins/src/main/java/org/gradle/initialization/buildsrc/GroovyBuildSrcProjectConfigurationAction.java:43]: This code is repeated, and should be abstracted/shared.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java [hunk 7]
```diff
@@ -416,6 +425,44 @@ protected ITaskFactory createITaskFactory(Instantiator instantiator, TaskClassIn
...
+                return version == null
+                    ? gradleApisFromCurrentGradle(dependencyFactory)
+                    : gradleApisFromRepository(resolutionServices, dependencyFactory, version);
+            },
+            () -> ((SelfResolvingDependency) dependencyFactory.createDependency(DependencyFactory.ClassPathNotation.LOCAL_GROOVY)).resolve());
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:451]: Looks like we are using local Groovy regardless of the Gradle version, but that will change. Shouldn't we resolve the Groovy version via the repository instead if the target API version is specified?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java [hunk 8]
```diff
@@ -266,6 +266,11 @@ public Dependency gradleApi() {
         return dependencyFactory.createDependency(DependencyFactory.ClassPathNotation.GRADLE_API);
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java:266]: Should we qualify the name of this constant now? Like, `CURRENT_GRADLE_API`? Or `BUNDLED_GRALDE_API`?

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt [hunk 9]
```diff
@@ -119,6 +119,7 @@ class KotlinBuildScriptIntegrationTest : AbstractKotlinIntegrationTest() {
             }
 
             task("test") {
+                doNotTrackState("no state")
```
Comment (lptr) [inline subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt:122]: Why is this necessary now?

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt [hunk 10]
```diff
@@ -71,6 +72,7 @@ fun fragmentsForConfiguration(accessor: Accessor.ForConfiguration): Fragments =
     val (functionFlags, deprecationBlock) =
         if (config.hasDeclarationDeprecations()) publicFunctionWithAnnotationsFlags to config.getDeclarationDeprecationBlock()
         else publicFunctionFlags to ""
+    val compileAgainstGradleVersion = GradleVersion.version(System.getProperty("org.gradle.api.version", GradleVersion.current().version)).baseVersion
```
Comment (lptr) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt:75]: I think we should call this "target Gradle version" everywhere.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyHandler.java:389]: Yeah, probably. Though I wouldn't go into polishing the Javadoc for this spike.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java (hunk 5 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java:165]: Yes, this can be improved. Note that with this solution you can already specify a different version of the gradle API by adding a constraint to the configuration, since the `gradleApi(version)` is a regular dependency and participates in dependency resolution. For the final solution, I am not sure what the right thing from dependency management is. `defaultDependencies` sounds it may be something in the right direction.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java (hunk 7 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:451]: I wouldn't resolve a different Groovy/Kotlin version per API. If we'd do this, then we'd also need to upgrade the Kotlin/Groovy code and use a different compiler as well. I don't see why we can't compile the old code with the newer Kotlin/Groovy version.

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt (hunk 9 above)
Comment (wolfs) [reply subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt:122]: This is my guinea pig for trying out if source pinning works, since the compilation will fail if I use a Gradle version older than 7.3.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:446]: Yeah, I think we should qualify the property name. It feels like `target` doesn't really capture it. For me, `target` sounds more about the target runtime. How about `org.gradle.api.source-version`, since this is the version of the Gradle API we use for sources.

File: subprojects/core/src/integTest/groovy/org/gradle/groovy/scripts/GroovyImportsIntegrationTest.groovy [hunk 11]
```diff
@@ -39,4 +40,19 @@ import org.apache.commons.math3.util.FastMath
         succeeds("help")
     }
 
+    @IgnoreRest
```
Comment (asodja) [inline subprojects/core/src/integTest/groovy/org/gradle/groovy/scripts/GroovyImportsIntegrationTest.groovy:43]: Is `@IgnoreRest` here for prototyping reasons?

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java]: Renamed.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java (hunk 4 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java]: I extracted the common logic to a single class. I didn't make it a service, yet, since I am not sure what scope this would need to go.

File: subprojects/plugins/src/main/java/org/gradle/initialization/buildsrc/GroovyBuildSrcProjectConfigurationAction.java (hunk 6 above)
Comment (wolfs) [reply subprojects/plugins/src/main/java/org/gradle/initialization/buildsrc/GroovyBuildSrcProjectConfigurationAction.java]: I extracted the common code.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java (hunk 8 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java:266]: I wouldn't do this now. I would like to get rid of this notation and always use a real dependency.

File: subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt (hunk 9 above)
Comment (wolfs) [reply subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt]: I removed the changes to the tests.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt (hunk 10 above)
Comment (wolfs) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt]: Renamed.

File: subprojects/core/src/integTest/groovy/org/gradle/groovy/scripts/GroovyImportsIntegrationTest.groovy (hunk 11 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/groovy/scripts/GroovyImportsIntegrationTest.groovy]: Yeah, I removed it now.

**Review by asodja — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClasspathProvider.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                private URI toUri(URL url) {
+                    try {
+                        return url.toURI();
+                    } catch (URISyntaxException e) {
+                        try {
+                            return new URL(
+                                url.getProtocol(),
+                                url.getHost(),
+                                url.getPort(),
+                                url.getFile().replace(" ", "%20")
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClasspathProvider.java:110]: I copied that from the Kotlin version of this class which seems to work. I suppose we should have only one classpath provider and not two separate implementations doing mostly the same things.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java (hunk 7 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:444]: I think that reasoning makes sense. Let's mention it here in a comment.
